File: swift/Tool.swift
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// A protocol that defines what types can be used as parameters in a LiteRT-LM tool.
public protocol ToolParameterValue: Codable {
  /// Returns the JSON schema representation for this type.
  static func getJsonSchema() -> [String: Any]
}

extension String: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] { ["type": "string"] }
}
extension Int: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] { ["type": "integer"] }
}
extension Bool: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] { ["type": "boolean"] }
}
extension Double: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] { ["type": "number"] }
}
extension Float: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] { ["type": "number"] }
}
extension Optional: ToolParameterValue where Wrapped: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] {
    var schema = Wrapped.getJsonSchema()
    schema["nullable"] = true
    return schema
  }
}
extension Array: ToolParameterValue where Element: ToolParameterValue {
  public static func getJsonSchema() -> [String: Any] {
    ["type": "array", "items": Element.getJsonSchema()]
  }
}

/// A non-generic protocol allowing `ToolManager` to inspect `ToolParam` properties at runtime.
protocol ToolParamProtocol {
  var description: String { get }
  var isRequired: Bool { get }

  /// Type-erased access to the underlying storage (default value), used for reflection.
  /// This allows `ToolManager` to read default values without knowing the specific generic `Value` type.
  var wrappedValueAny: Any? { get }

  func getFullSchema() -> [String: Any]
}

/// A property wrapper that defines a parameter for a LiteRT-LM tool.
///
/// Use this property wrapper to mark properties of a `Tool` as parameters that can be passed to
/// the tool. The `description` will be used to generate the tool's schema.
///
/// Example:
/// ```
/// // Example of required parameters:
/// @ToolParam(description: "The title of the event.")
/// var title: String
/// @ToolParam(description: "The date and time of the event in format YYYY-MM-DDTHH:MM:SS.")
/// var dateTime: String
///
/// // Example of a parameter with a default value:
/// @ToolParam(description: "The duration of the event in minutes. Default is 30.")
/// var durationMinutes: Int = 30
///
/// // Example of an optional parameter:
/// @ToolParam(description: "The topic of the event.")
/// var topic: String?
/// ```
@propertyWrapper
public struct ToolParam<Value: ToolParameterValue>: Decodable, ToolParamProtocol {

  /// Holds the actual value of the parameter.
  private var storage: Value?

  /// A description of the parameter. This is used to to describe the parameter in the schema.
  public let description: String

  /// Whether the parameter has a default value provided by the developer when declaring the tool.
  private let hasDefaultValue: Bool

  /// Initializes a parameter with an optional default value.
  public init(wrappedValue: Value? = nil, description: String) {
    self.storage = wrappedValue
    self.description = description
    self.hasDefaultValue = wrappedValue != nil
  }

  /// The value of the parameter.
  ///
  /// Returns the stored value if available (from a default or decoded input).
  /// Returns `nil` if no value is present and the type `Value` is Optional.
  public var wrappedValue: Value {
    get {
      // If we have data, return it.
      if let value = storage {
        return value
      }

      // If storage is empty, check if it's an Optional.
      if let nilValue = Optional<Any>.none as? Value {
        return nilValue
      }

      // It's not optional and we have no data. Throw an error.
      fatalError("ToolParam of type \(Value.self) was not set and has no default value.")
    }
    set { storage = newValue }
  }

  /// Provides type-erased access to the underlying storage of the parameter.
  ///
  /// Unlike `wrappedValue`, this property returns `nil` if no value is set, instead of crashing.
  /// This is essential for `ToolManager` to safely inspect default values via reflection (Mirror)
  /// before the decoding process occurs.
  var wrappedValueAny: Any? {
    return storage
  }

  func getFullSchema() -> [String: Any] {
    var schema = Value.getJsonSchema()
    if !description.isEmpty {
      schema["description"] = description
    }
    return schema
  }

  var isRequired: Bool {
    // If `nil` can be cast to `Value`, then `Value` must be an Optional type.
    let isOptional = Optional<Any>.none as? Value != nil
    return !isOptional && !hasDefaultValue
  }

  /// Decodes the parameter value from the decoder.
  ///
  /// This initializer is required to conform to the `Decodable` protocol. Since the `Tool` protocol
  /// inherits from `Decodable`, any property wrapper used within a `Tool` must also be `Decodable`.
  ///
  /// When `ToolManager` executes a tool, it uses `JSONDecoder` to decode the JSON arguments
  /// provided by the model into a `Tool` instance. This initializer is called automatically during
  /// that process to populate the `ToolParam`'s internal storage with the value provided by the
  /// model.
  public init(from decoder: Decoder) throws {
    let container = try decoder.singleValueContainer()

    if container.decodeNil() {
      if let nilValue = Optional<Any>.none as? Value {
        self.storage = nilValue
      } else {
        throw DecodingError.valueNotFound(
          Value.self,
          .init(
            codingPath: decoder.codingPath,
            debugDescription: "Received null for non-optional parameter"))
      }
    } else {
      self.storage = try container.decode(Value.self)
    }

    self.description = ""
    self.hasDefaultValue = true
  }
}

/// Example of how to define tools:
/// - Adopt the `Tool` protocol to define a struct/class as a tool.
/// - Use `@ToolParam` to define the parameters of the tool.
/// - The allowed parameter types are: String, Int, Bool, Float, Double, and Array of them.
/// - The return type of `run()` is `Any` and will be converted to JSON/String back to the model.
/// - Use the Swift Optional type (e.g., `String?`) to indicate that a parameter is optional.
///
/// ```swift
/// struct GetCurrentWeatherTool: Tool {
///   static let name = "get_current_weather"
///   static let description = "Get the current weather"
///
///   @ToolParam(description: "The city and state, e.g. San Francisco, CA")
///   var location: String
///
///   @ToolParam(description: "The temperature unit to use")
///   var unit: String = "celsius"
///
///   func run() async throws -> Any {
///     // In a real application, you would call a weather API here. This is just an example.
///     let temperature = getTemperature(for: location, in: unit)
///
///     return [
///       "temperature": temperature,
///       "unit": unit,
///     ]
///   }
/// }
/// ```
///  A protocol a struct/class must conform to be used as a tool by the LiteRT-LM model.
///
///  - SeeAlso: `ToolParam` for defining parameters of the tool.
public protocol Tool: Decodable {
  /// The unique name of the tool.
  static var name: String { get }
  /// A description of what the tool does.
  static var description: String { get }

  /// A required zero-argument initializer.
  init()
  /// The logic to execute when the tool is called.
  func run() async throws -> Any
}

extension Tool {
  /// Generates an OpenAPI-compliant JSON schema for the tool.
  public func getSchema() -> [String: Any] {
    var properties: [String: Any] = [:]
    var requiredFields: [String] = []

    let mirror = Mirror(reflecting: self)
    let useSnakeCase = ExperimentalFlags.convertCamelToSnakeCaseInToolDescription

    for child in mirror.children {
      guard let label = child.label else { continue }

      // Clean the name (remove the _ prefix and handle snake_case)
      let cleanName = label.cleanPropertyLabel(useSnakeCase: useSnakeCase)

      // Check if this property is a ToolParam
      if let param = child.value as? ToolParamProtocol {
        properties[cleanName] = param.getFullSchema()
        if param.isRequired {
          requiredFields.append(cleanName)
        }
      }
    }

    let toolName = useSnakeCase ? Self.name.camelToSnakeCase() : Self.name

    var parameters: [String: Any] = [
      "type": "object",
      "properties": properties,
    ]

    if !requiredFields.isEmpty {
      parameters["required"] = requiredFields
    }

    var schema: [String: Any] = [
      "type": "function",
      "function": [
        "name": toolName,
        "description": Self.description,
      ],
    ]

    if !properties.isEmpty {
      var functionBody = schema["function"] as! [String: Any]
      functionBody["parameters"] = parameters
      schema["function"] = functionBody
    }

    return schema
  }
}

extension String {
  /// Converts a camelCase string to snake_case.
  func camelToSnakeCase() -> String {
    let regex = try! NSRegularExpression(pattern: "([a-z0-9])([A-Z])", options: [])
    let range = NSRange(location: 0, length: self.utf16.count)
    return regex.stringByReplacingMatches(
      in: self, options: [], range: range, withTemplate: "$1_$2"
    ).lowercased()
  }

  /// Cleans a property label (removes '_' prefix) and optionally converts to snake_case.
  func cleanPropertyLabel(useSnakeCase: Bool) -> String {
    let cleaned = self.hasPrefix("_") ? String(self.dropFirst()) : self
    return useSnakeCase ? cleaned.camelToSnakeCase() : cleaned
  }
}

File: kotlin/java/com/google/ai/edge/litertlm/Tool.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.ai.edge.litertlm

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import com.google.gson.JsonPrimitive
import kotlin.reflect.KParameter
import kotlin.reflect.full.functions

/**
 * Example of how to define tools with `AutoToolSet`:
 * - Use `@Tool` to define a method as a tool.
 * - Use `@ToolParam` to add information to the param of a tool.
 * - The allowed parameter types are: String, Int, Boolean, Float, Double, and List of them.
 * - The return type of your tool can be any Kotlin type.
 *
 * ```kotlin
 * class MyToolSet: AutoToolSet() {
 *   @Tool(description = "Get the current weather")
 *   fun getCurrentWeather(
 *     @ToolParam(description = "The city and state, e.g. San Francisco, CA") location: String,
 *     @ToolParam(description = "The temperature unit to use") unit: String = "celsius",
 *   ): Map {
 *     return mapOf(
 *       "temperature" to 25,
 *       "unit" to "Celsius",
 *     )
 *   }
 * }
 * ```
 */

/**
 * Annotation to mark a function as a tool that can be used by the LiteRT-LM model.
 *
 * @property description A description of the tool.
 */
@Target(AnnotationTarget.FUNCTION) // This annotation can only be applied to functions
@Retention(AnnotationRetention.RUNTIME) // IMPORTANT: Makes the annotation available at runtime
annotation class Tool(val description: String)

/**
 * Annotation to provide a description for a tool parameter.
 *
 * @property description A description of the tool parameter.
 */
@Target(AnnotationTarget.VALUE_PARAMETER) // This annotation can only be applied to functions
@Retention(AnnotationRetention.RUNTIME) // IMPORTANT: Makes the annotation available at runtime
annotation class ToolParam(val description: String)

/**
 * Tool provider for Conversation.
 *
 * Use [tool] function to convert the supported tool types to this [ToolProvider].
 */
abstract class ToolProvider {

  /**
   * Provides a map of tools, where the key is the tool name and the value is the tool
   * implementation.
   *
   * This method uses reflection to find all functions annotated with @Tool in the current class. It
   * then wraps each function in a ReflectionTool instance and returns them in a map.
   */
  internal abstract fun provideTools(): Map<String, InternalJsonTool>
}

/**
 * An abstract class for defining collection of tools using Kotlin functions.
 * - Use `@Tool` to define a method as a tool.
 * - Use `@ToolParam` to add information to the param of a tool.
 * - The allowed parameter types are: String, Int, Boolean, Float, Double, and List of them.
 * - The return type of your tool can be any Kotlin type.
 *
 * ```kotlin
 * class MyToolSet: ToolSet {
 *   @Tool(description = "Get the current weather")
 *   fun getCurrentWeather(
 *     @ToolParam(description = "The city and state, e.g. San Francisco, CA") location: String,
 *     @ToolParam(description = "The temperature unit to use") unit: String = "celsius",
 *   ): Map {
 *     return mapOf(
 *       "temperature" to 25,
 *       "unit" to "Celsius",
 *     )
 *   }
 * }
 * ```
 */
interface ToolSet {}

/**
 * An abstract class for defining a tool using an Open API specification.
 *
 * This class provides a way to define a tool by providing a JSON string that conforms to the Open
 * API specification. The tool can then be executed by providing a JSON string containing the
 * parameters.
 *
 * The input and output are both JSON strings to provide flexibility in choosing a JSON library.
 * This avoids forcing a specific library on the user, allowing them to use the one that best suits
 * their project needs and dependencies.
 */
interface OpenApiTool {

  /**
   * Gets the tool description JSON string based on the Open API specification.
   *
   * The JSON string should be a valid JSON object with keys:
   * - name : Required. The name of the tool.
   * - description: Optional. A brief description of the function.
   * - parameters: Optional. Describes the parameters to this function.
   *
   * For example,
   * - {"name":"addition","description":"Add all
   *   numbers.","parameters":{"type":"object","properties":{"numbers":{"type":"array","items":{"type":"number"}},"description":"The
   *   list of numbers to sum."},"required":["numbers"]}}
   */
  fun getToolDescriptionJsonString(): String

  /**
   * Executes the tools with the paramsJsonString and return the result as string.
   *
   * The paramsJsonString is a JSONObject where the keys are the name of the parameters.
   *
   * For examples, {"numbers":[3.0,4.5,6.0]}
   *
   * The return value is the JSON string of the results. For examples,
   * - 13.5 // return a JSON primitive value
   * - {"sum": 13.5} // return as a JSON Object
   */
  fun execute(paramsJsonString: String): String
}

/** Returns a [ToolProvider] from a [ToolSet]. */
fun tool(toolSet: ToolSet): ToolProvider {
  return object : ToolProvider() {
    @OptIn(ExperimentalApi::class)
    override fun provideTools(): Map<String, InternalJsonTool> {
      val useSnakeCase = ExperimentalFlags.convertCamelToSnakeCaseInToolDescription

      val toolClass = toolSet.javaClass.kotlin
      return toolClass.functions
        .filter { function -> function.annotations.any { annotation -> annotation is Tool } }
        .map { function ->
          (if (useSnakeCase) function.name.camelToSnakeCase() else function.name) to
            ReflectionTool(toolSet, function, useSnakeCase) // Pass toolSet here
        }
        .toMap()
    }
  }
}

/** Returns a [ToolProvider] from an [OpenApiTool]. */
fun tool(openApiTool: OpenApiTool): ToolProvider {
  return object : ToolProvider() {
    override fun provideTools(): Map<String, InternalJsonTool> {
      val toolDescription: JsonObject =
        try {
          JsonParser.parseString(openApiTool.getToolDescriptionJsonString()).asJsonObject
        } catch (e: JsonParseException) {
          throw ToolException("Failed to parse JSON. ${e.message}", e)
        }

      val name: String =
        try {
          toolDescription.get("name").asString
        } catch (e: Throwable) {
          throw ToolException("Failed to parse field \"name\" as String. ${e.message}", e)
        }

      val jsonTool =
        object : InternalJsonTool {
          override fun getToolDescription(): JsonObject {
            return toolDescription
          }

          override fun execute(params: JsonObject): Any? {
            return openApiTool.execute(params.toString())
          }
        }

      return mapOf(name to jsonTool)
    }
  }
}

/**
 * Manages a collection of tools and provides methods to execute tools and get their specifications.
 *
 * @property tools A list of [ToolProvider].
 */
class ToolManager(val tools: List<ToolProvider> = emptyList()) {

  private val internalTools: Map<String, InternalJsonTool> =
    tools.fold(mapOf()) { acc, tool -> acc + tool.provideTools() }

  /**
   * Executes a tool function by its name with the given parameters.
   *
   * @param functionName The name of the tool function to execute.
   * @param params A JsonObject containing the parameter names and their values.
   * @return The result of the tool function execution as a string.
   * @throws IllegalArgumentException if the tool function is not found.
   */
  fun execute(functionName: String, params: JsonObject): JsonElement {
    try {
      val tool =
        internalTools[functionName]
          ?: throw IllegalArgumentException("Tool not found: ${functionName}")
      return tool.execute(params).toJsonElement()
    } catch (e: Exception) {
      return JsonPrimitive("Error occured. ${e.toString()}")
    }
  }

  /**
   * Gets the tools description for all registered tools in Open API format.
   *
   * @return A json array of OpenAPI tool description JSON as string.
   */
  fun getToolsDescription() =
    JsonArray().apply {
      for (tool in internalTools.values) {
        // Wrap the Open API spec in function object, expected by the native library.
        add(
          JsonObject().apply {
            addProperty("type", "function")
            add("function", tool.getToolDescription())
          }
        )
      }
    }
}

/** Internal use only. */
interface InternalJsonTool {

  fun getToolDescription(): JsonObject

  fun execute(params: JsonObject): Any?
}

/**
 * Represents a single tool, wrapping an instance and a specific Kotlin function.
 *
 * @property instance The instance of the class containing the tool function.
 * @property kFunction The Kotlin function to be executed as a tool.
 * @property useSnakeCase Whether to use snake case for function and param names for tool calling.
 */
internal class ReflectionTool(
  val instance: Any,
  val kFunction: kotlin.reflect.KFunction<*>,
  val useSnakeCase: Boolean,
) : InternalJsonTool {

  /**
   * Gets the tool description in Open API format.
   *
   * @return The tool description.
   */
  override fun getToolDescription(): JsonObject {
    val toolAnnotation = kFunction.annotations.find { it is Tool } as? Tool ?: return JsonObject()

    val description = toolAnnotation.description

    val openApiSpec =
      JsonObject().apply {
        val funcName = if (useSnakeCase) kFunction.name.camelToSnakeCase() else kFunction.name
        addProperty("name", funcName)
        addProperty("description", description)
      }

    val parameters = kFunction.parameters.drop(1) // Drop the instance parameter
    if (!parameters.isEmpty()) {
      val properties = JsonObject()
      for (param in parameters) {
        val paramAnnotation = param.annotations.find { it is ToolParam } as? ToolParam
        val paramJsonSchema = getTypeJsonSchema(param.type)
        // add "description" if provided
        paramAnnotation?.description?.let { paramJsonSchema.addProperty("description", it) }
        if (param.type.isMarkedNullable) paramJsonSchema.addProperty("nullable", true)
        properties.add(param.toModelParamName(), paramJsonSchema)
      }

      val requiredParams = JsonArray()
      for (param in parameters) {
        if (!param.isOptional) {
          requiredParams.add(param.toModelParamName())
        }
      }

      val schema =
        JsonObject().apply {
          addProperty("type", "object")
          add("properties", properties)
          if (!requiredParams.isEmpty) add("required", requiredParams)
        }

      openApiSpec.add("parameters", schema)
    }

    return openApiSpec
  }

  /**
   * Executes the tool function with the given parameters.
   *
   * @param params A JsonObject containing the parameter names and their values.
   * @return The result of the tool function execution as a Any?.
   * @throws IllegalArgumentException if any required parameters are missing.
   */
  override fun execute(params: JsonObject): Any? {
    val args =
      kFunction.parameters
        .associateWith { param ->
          when {
            param.index == 0 -> instance // First parameter is the instance
            param.name != null && params.has(param.toModelParamName()) -> {
              val value = params.get(param.toModelParamName())
              convertJsonValueToKotlinValue(value, param.type)
            }
            param.isOptional -> null // Should not be reached
            else -> throw IllegalArgumentException("Missing parameter: ${param.toModelParamName()}")
          }
        }
        .filterValues { it != null }

    return kFunction.callBy(args)
  }

  /**
   * Converts a JSON value to the expected Kotlin type.
   *
   * @param value The JSON value to convert.
   * @param type The target Kotlin type.
   * @return The converted value.
   * @throws IllegalArgumentException if the value cannot be converted to the target type.
   */
  private fun convertJsonValueToKotlinValue(value: JsonElement, type: kotlin.reflect.KType): Any {
    val classifier = type.classifier
    return when {
      classifier == List::class && value is JsonArray -> {
        val listTypeArgument = type.arguments.firstOrNull()?.type
        value.map { convertJsonValueToKotlinValue(it, listTypeArgument!!) }
      }
      classifier == Int::class && value is JsonPrimitive && value.isNumber -> value.asInt
      classifier == Float::class && value is JsonPrimitive && value.isNumber -> value.asFloat
      classifier == Double::class && value is JsonPrimitive && value.isNumber -> value.asDouble
      classifier == String::class && value is JsonPrimitive && value.isString -> value.asString
      classifier == Boolean::class && value is JsonPrimitive && value.isBoolean -> value.asBoolean
      // Add more conversions if needed
      else -> value
    }
  }

  /**
   * Generates a JSON schema for the given Kotlin type.
   *
   * @param type The Kotlin type to generate the schema for.
   * @return A JsonObject representing the JSON schema.
   * @throws IllegalArgumentException if the type is not supported.
   */
  private fun getTypeJsonSchema(type: kotlin.reflect.KType): JsonObject {
    val classifier = type.classifier
    val jsonType = javaTypeToJsonTypeString[classifier]

    if (jsonType == null) {
      throw IllegalArgumentException(
        "Unsupported type: ${classifier.toString()}. " +
          "Allowed types are: ${javaTypeToJsonTypeString.keys.joinToString { it.simpleName ?: "" }}"
      )
    }

    val schema = JsonObject()
    schema.addProperty("type", jsonType)
    if (classifier == List::class) {
      val listTypeArgument = type.arguments.firstOrNull()?.type
      if (listTypeArgument == null) {
        throw IllegalArgumentException("List type argument is missing.")
      }
      schema.add("items", getTypeJsonSchema(listTypeArgument))
    }
    return schema
  }

  private fun KParameter.toModelParamName(): String {
    return if (useSnakeCase) this.name!!.camelToSnakeCase() else this.name!!
  }

  companion object {
    private val javaTypeToJsonTypeString =
      mapOf(
        String::class to "string",
        Int::class to "integer",
        Boolean::class to "boolean",
        Float::class to "number",
        Double::class to "number",
        List::class to "array",
      )
  }
}

private fun String.camelToSnakeCase(): String {
  return this.replace(Regex("(?<=[a-zA-Z])[A-Z]")) { "_${it.value}" }.lowercase()
}

private fun String.snakeToCamelCase(): String {
  return Regex("_([a-z])").replace(this) { it.value.substring(1).uppercase() }
}

/** Exception related to tool calling. */
class ToolException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)
